pkgs/os-specific/linux/zfs/default.nix

```diff
@@ -97,6 +103,36 @@ let
         substituteInPlace ./cmd/vdev_id/vdev_id \
           --replace "PATH=/bin:/sbin:/usr/bin:/usr/sbin" \
           "PATH=${makeBinPath [ coreutils gawk gnused gnugrep systemd ]}"
+      '' + optionalString (buildUser) ''
+        # TODO(ricsch): Is overriding with /run/current-system/sw/bin acceptable?
+        substituteInPlace scripts/zfs-tests.sh \
+          --replace '$STF_PATH/ksh' ${ksh}/bin/ksh \
+          --replace 'LOSETUP=' 'LOSETUP=${util-linux}/bin/losetup #' \
+          --replace 'export PATH=$STF_PATH' '#' \
+          --replace 'DMSETUP=' 'DMSETUP=${lvm2}/bin/dmsetup #' \
+          --replace '$STF_PATH/gzip' '${gzip}/bin/gzip' \
+          --replace '$STF_PATH/gunzip' '${gzip}/bin/gunzip' \
+          --replace '/sbin/fsck.ext4' '${e2fsprogs}/bin/fsck.ext4' \
+          --replace '/sbin/mkfs.ext4' '${e2fsprogs}/bin/mkfs.ext4' \
+          --replace '$STF_PATH/awk' '${gawk}/bin/awk' \
+          --replace 'SYSTEM_DIRS="/usr/local/bin /usr/local/sbin"' 'SYSTEM_DIRS="/run/current-system/sw/bin"'
+
+        substituteInPlace scripts/common.sh.in \
+          --replace 'export ZTS_DIR=' 'export ZTS_DIR=${placeholder "zfs_tests"}/share/zfs #' \
+          --replace 'export SCRIPT_DIR=' 'export SCRIPT_DIR=${placeholder "zfs_tests"}/share/zfs #' \
+          --replace 'export ZDB=' 'export ZDB=${placeholder "out"}/bin/zdb #' \
+          --replace 'export ZFS=' 'export ZFS=${placeholder "out"}/bin/zfs #' \
+          --replace 'export ZPOOL=' 'export ZPOOL=${placeholder "out"}/bin/zpool #' \
+          --replace 'export ZTEST=' 'export ZTEST=${placeholder "out"}/bin/ztest #' \
+          --replace 'export ZFS_SH=' 'export ZFS_SH=${placeholder "zfs_tests"}/share/zfs/zfs.sh #'
+
+        # Fix ksh paths in test suite.
+        # patchShebangs doesn't work due to the scripts not being executable.
+        # It doesn't seem logical to make them executable either.
+        find -name "*.ksh" -exec sed -i 's,/bin/ksh,${ksh}/bin/ksh,' {} \;
+
+        # Patching maybe required for more binaries
+        # Maybe FHS environment would be better?
```

Comment:
Oh boy, look at the time! It's time to resurrect this PR yet again :)

I think these scripts could use [`makeWrapper`](https://nixos.org/manual/nixpkgs/stable/#fun-makeWrapper) instead of awkwardly patching them. Most variables use the pattern `LOSETUP=${LOSETUP:-/sbin/losetup}`, so they can be overridden using `makeWrapper ... --set LOSETUP "${util-linux}/bin/losetup"`.

Also what's up with all the ksh stuff? I've read through `zfs-tests.sh` and to me it looks like ksh is only used on FreeBSD.

Reply:
I agree! Hopefully I can find some time soon to maybe clean this up :). Of course, feel free to contribute if you feel like it.